eth/dao_hard_fork.go

```diff
@@ -0,0 +1,197 @@
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+// Package eth implements the Ethereum protocol.
+package eth
+
+// WARNING: THIS CODE IS WORK IN PROGRESS AND NEEDS EXTENSIVE REVIEW
+//
+// Utility functions for preparing and enacting a hard fork to revert the hack of "The DAO"
+// according to proposed spec at:
+// https://docs.google.com/document/d/1VfuAH7Zf0UQmuVw1o7cTPbNtp1wFYzlN1tBKV6SbVSI
+//
+
+import (
+	"encoding/json"
+	"math/big"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/internal/dao"
+	"github.com/ethereum/go-ethereum/logger"
+	"github.com/ethereum/go-ethereum/logger/glog"
+)
+
+var (
+	theDAOAddr = common.HexToAddress("bb9bc244d798123fde783fcc1c72d3bb8c189413")
+	// TODO: set these after community consensus
+	theDAOReplacementCode = []byte{0x00}
+	hardForkBlock         = big.NewInt(999999999)
+)
+
+type DAOHardForkData struct {
+	ChildDAOs             []common.Address
+	ExtraBalances         []common.Address
+	TheDAOReplacementCode []byte
+}
+
+func TestRunHardFork(eth *Ethereum) error {
+	hfData, err := prepareHardForkData(eth)
+	if err != nil {
+		return err
+	}
+
+	// log JSON representation of hard fork data for manual inspection
+	logHardForkData(hfData, eth)
+
+	// uncomment to (for testing) apply hard fork on current state:
+	// err = applyHardFork(hfData, eth)
+	return err
+}
+
+func logHardForkData(hfData *DAOHardForkData, eth *Ethereum) {
+	j, _ := json.Marshal(hfData)
+	glog.V(logger.Info).Infof("Generated hard fork data: %s", j)
+
+	statedb, err := eth.BlockChain().State()
+	if err != nil {
+		return
+	}
+
+	// only accumulated for logging purposes
+	total := big.NewInt(0)
+	for i := 0; i < len(hfData.ChildDAOs); i++ {
+		acc := statedb.GetAccount(hfData.ChildDAOs[i])
+		total.Add(total, acc.Balance())
+	}
+
+	for i := 0; i < len(hfData.ExtraBalances); i++ {
+		acc := statedb.GetAccount(hfData.ExtraBalances[i])
+		total.Add(total, acc.Balance())
+	}
+
+	glog.V(logger.Info).Infof("Total wei in all accounts: %v", total)
+}
+
+func prepareHardForkData(eth *Ethereum) (*DAOHardForkData, error) {
+	hfData := &DAOHardForkData{
+		TheDAOReplacementCode: theDAOReplacementCode,
+	}
+
+	theDAO, err := dao.NewDAO(theDAOAddr, NewContractBackend(eth))
+	if err != nil {
+		glog.V(logger.Info).Infof("could not instantiate DAO: %v", err)
+		return nil, err
+	}
+
+	// create list of "The DAO" and all its children, recursively
+	daos, err := traverseDAOs(theDAO, eth)
```

Comment:
this is the intention - the code comments need an update here though - good catch.
the original DAO is excluded as it has 1 wei, it's the ExtraBalance at 0x807640a13483f8ac783c557fcdf27be11ea4ac7a does have 345K ETH so that should probably be included in the spec (unless the replacement code for 0xbb9bc244d798123fde783fcc1c72d3bb8c189413 would include code to tap that account).

The spec should be clarified here (see linked google doc) - feel free to comment there too.